### ISSUE TITLE
Illegible error message if sassc is missing

### ISSUE THREAD

**Issue by martinl88** [cockpit-project/starter-kit#391]
Running the example...
```
git clone https://github.com/cockpit-project/starter-kit.git
cd starter-kit
make
```

...ends with:
```
npm ERR! code ELIFECYCLE
npm ERR! errno 2
npm ERR! starter-kit@0.1.0 build: 'webpack'
npm ERR! Exit status 2
npm ERR!
npm ERR! Failed at the starter-kit@0.1.0 build script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
npm ERR! A complete log of this run can be found in:
npm ERR!     /home/dev/.npm/_logs/2020-11-05T14_40_15_913Z-debug.log
make: *** [dist/index.css] Error 2
```
**2020-11-05T14_40_15_913Z-debug.log**
```
0 info it worked if it ends with ok
1 verbose cli [ '/usr/bin/node', '/usr/bin/npm', 'run', 'build' ]
2 info using npm@6.14.8
3 info using node@v12.19.0
4 verbose run-script [ 'prebuild', 'build', 'postbuild' ]
5 info lifecycle starter-kit@0.1.0~prebuild: starter-kit@0.1.0
6 info lifecycle starter-kit@0.1.0~build: starter-kit@0.1.0
7 verbose lifecycle starter-kit@0.1.0~build: unsafe-perm in lifecycle true
8 verbose lifecycle starter-kit@0.1.0~build: PATH: /usr/lib/node_modules/npm/node_modules/npm-lifecycle/node-gyp-bin:/home/dev/starter-kit/node_modules/.bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/home/dev/bin
9 verbose lifecycle starter-kit@0.1.0~build: CWD: /home/dev/starter-kit
10 silly lifecycle starter-kit@0.1.0~build: Args: [ '-c', 'webpack' ]
11 silly lifecycle starter-kit@0.1.0~build: Returned: code: 2  signal: null
12 info lifecycle starter-kit@0.1.0~build: Failed to exec build script
13 verbose stack Error: starter-kit@0.1.0 build: `webpack`
13 verbose stack Exit status 2
13 verbose stack     at EventEmitter.<anonymous> (/usr/lib/node_modules/npm/node_modules/npm-lifecycle/index.js:332:16)
13 verbose stack     at EventEmitter.emit (events.js:314:20)
13 verbose stack     at ChildProcess.<anonymous> (/usr/lib/node_modules/npm/node_modules/npm-lifecycle/lib/spawn.js:55:14)
13 verbose stack     at ChildProcess.emit (events.js:314:20)
13 verbose stack     at maybeClose (internal/child_process.js:1021:16)
13 verbose stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:286:5)
14 verbose pkgid starter-kit@0.1.0
15 verbose cwd /home/dev/starter-kit
16 verbose Linux 3.10.0-1127.el7.x86_64
17 verbose argv "/usr/bin/node" "/usr/bin/npm" "run" "build"
18 verbose node v12.19.0
19 verbose npm  v6.14.8
20 error code ELIFECYCLE
21 error errno 2
22 error starter-kit@0.1.0 build: `webpack`
22 error Exit status 2
23 error Failed at the starter-kit@0.1.0 build script.
23 error This is probably not a problem with npm. There is likely additional logging output above.
24 verbose exit [ 2, true ]
```

**Versions:**
```
[root@localhost starter-kit]# node -v
v12.19.0
[root@localhost starter-kit]# npm -v
6.14.8
[root@localhost starter-kit]# webpack -v
5.4.0
```

What am I missing?

**Comment by martinpitt** [comment]
We don't support webpack 5 yet. Please uninstall webpack, or build this in an environment without pre-installed webpack. starter-kit will install webpack through NPM, but apparently it prefers the system-installed version.

**Comment by martinl88** [comment]
OK.
To be absolutely sure, I deployed a fresh CentOS 7 machine and tried again... but ended up in exactly the same place.

```
npm ERR! code ELIFECYCLE
npm ERR! errno 2
npm ERR! starter-kit@0.1.0 build: `webpack`
npm ERR! Exit status 2
npm ERR!
npm ERR! Failed at the starter-kit@0.1.0 build script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /home/devel/.npm/_logs/2020-11-05T15_22_00_059Z-debug.log
make: *** [dist/index.css] Error 2
```

**2020-11-05T15_22_00_059Z-debug.log**
```
0 info it worked if it ends with ok
1 verbose cli [ '/usr/bin/node', '/bin/npm', 'run', 'build' ]
2 info using npm@6.14.8
3 info using node@v12.19.0
4 verbose run-script [ 'prebuild', 'build', 'postbuild' ]
5 info lifecycle starter-kit@0.1.0~prebuild: starter-kit@0.1.0
6 info lifecycle starter-kit@0.1.0~build: starter-kit@0.1.0
7 verbose lifecycle starter-kit@0.1.0~build: unsafe-perm in lifecycle true
8 verbose lifecycle starter-kit@0.1.0~build: PATH: /usr/lib/node_modules/npm/node_modules/npm-lifecycle/node-gyp-bin:/home/devel/starter-kit/node_modules/.bin:/usr/local/bin:/bin:/usr/bin:/usr/local/sbin:/usr/sbin:/home/devel/.local/bin:/home/devel/bin
9 verbose lifecycle starter-kit@0.1.0~build: CWD: /home/devel/starter-kit
10 silly lifecycle starter-kit@0.1.0~build: Args: [ '-c', 'webpack' ]
11 silly lifecycle starter-kit@0.1.0~build: Returned: code: 2  signal: null
12 info lifecycle starter-kit@0.1.0~build: Failed to exec build script
13 verbose stack Error: starter-kit@0.1.0 build: `webpack`
13 verbose stack Exit status 2
13 verbose stack     at EventEmitter.<anonymous> (/usr/lib/node_modules/npm/node_modules/npm-lifecycle/index.js:332:16)
13 verbose stack     at EventEmitter.emit (events.js:314:20)
13 verbose stack     at ChildProcess.<anonymous> (/usr/lib/node_modules/npm/node_modules/npm-lifecycle/lib/spawn.js:55:14)
13 verbose stack     at ChildProcess.emit (events.js:314:20)
13 verbose stack     at maybeClose (internal/child_process.js:1021:16)
13 verbose stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:286:5)
14 verbose pkgid starter-kit@0.1.0
15 verbose cwd /home/devel/starter-kit
16 verbose Linux 3.10.0-1127.el7.x86_64
17 verbose argv "/usr/bin/node" "/bin/npm" "run" "build"
18 verbose node v12.19.0
19 verbose npm  v6.14.8
20 error code ELIFECYCLE
21 error errno 2
22 error starter-kit@0.1.0 build: `webpack`
22 error Exit status 2
23 error Failed at the starter-kit@0.1.0 build script.
23 error This is probably not a problem with npm. There is likely additional logging output above.
24 verbose exit [ 2, true ]
```

Webpack is at version 4 now.
```
[devel@localhost starter-kit]$ node -v
v12.19.0
[devel@localhost starter-kit]$ npm -v
6.14.8
[devel@localhost starter-kit]$ npm list webpack
starter-kit@0.1.0 /home/devel/starter-kit
└── webpack@4.44.2
```

**Comment by martinpitt** [comment]
Hmm, in most cases this happens when a module on npmjs.org gets updated and introduces a regression. But I just ran the same commands on my system, i.e. a clean tree, and it still builds fine. But let's compare the generated package-lock.json? I put mine [here](https://piware.de/tmp/package-lock.json). Feel free to diff it yourself, or just attach your's here or in a pastebin or so. Thanks!

**Comment by martinpitt** [comment]
verbose lifecycle starter-kit@0.1.0~build: unsafe-perm in lifecycle true

I wonder if that's relevant or a red herring. I don't see any of these verbose messages when running `make`. But it does happen very close to the failure, so it seems relevant?

    PATH: /usr/lib/node_modules/npm/node_modules/npm-lifecycle/node-gyp-bin:/home/devel/starter-kit/node_modules/.bin:/usr/local/bin:/bin:/usr/bin:/usr/local/sbin:/usr/sbin:/home/devel/.local/bin:/home/devel/bin

Could you perhaps try to run this without your $HOME parts? something like

    PATH=/usr/local/bin:/bin:/usr/bin:/usr/local/sbin:/usr/sbin make

If that works, put the others back one by one to see which path is the problematic one, and `ls -ld <path>` that? Thanks!

**Comment by martinl88** [comment]
My package-lock.json is identical to yours.

Tried cutting the PATH shorter but sadly still no dice :(
[2020-11-06T09_05_58_164Z-debug.log](https://github.com/cockpit-project/starter-kit/files/5500361/2020-11-06T09_05_58_164Z-debug.log) 
[cockpit-make-log.txt](https://github.com/cockpit-project/starter-kit/files/5500362/cockpit-make-log.txt)

Regarding the "unsafe-perm in lifecycle true" -> https://github.com/nodejs/node/issues/4943#issuecomment-623135889
It makes sense, as my latest test-environment is also running in non-root mode

I'm sure, I'm doing something fundamentally wrong here, but I just can figure out what that might be.

**Comment by martinl88** [comment]
[cockpit-make-log-2.txt](https://github.com/cockpit-project/starter-kit/files/5502171/cockpit-make-log-2.txt)

**Comment by martinl88** [comment]
Would it be possible for you to do a test setup on a completely fresh install?
I'm running out of ideas on what to try next. Is it possible that you have some additional env. variables set that I don't have?

**Comment by martinpitt** [comment]
I'm trying this in a completely clean environment, a current Fedora 33 container:
```
❱❱❱ podman run -it registry.fedoraproject.org/fedora:33 bash
# dnf install -y npm git make sassc
# useradd user
# su - user
$ git clone https://github.com/cockpit-project/starter-kit.git
$ cd starter-kit
$ make
```
which works just fine. Version comparison:
```
$ npm -v
6.14.8
$ node -v
v14.14.0
$ npm list webpack
starter-kit@0.1.0 /home/user/starter-kit
└── webpack@4.44.2 
```
So npn and webpack are the same; F33 has a newer node, yours was 12.19. That version is still in Fedora 31, so I retried the above in `registry.fedoraproject.org/fedora:31`, and that worked just as well.

You mentioned CentOS 7 above -- but that has neither npm nor sassc. RHEL/CentOS isn't really a good developer platform, you'd need additional repositories. Can you describe precisely how you set up your environment? And perhaps try Fedora, Debian, or Ubuntu?

**Comment by martinpitt** [comment]
Oh! I just looked at your cockpit-make-log-2.txt and that is *not* the same error any more -- it just says that the `sassc` program was not found. Please install the "sassc" package, and you should be good.

**Comment by martinl88** [comment]
Hooray! Missing **sassc** was indeed the culprit. After installing that, compile is successful!
Thank you so much. I still don't understand how I managed to miss those errors.

**Comment by martinpitt** [comment]
Reopening. This was [reported again](https://lists.fedorahosted.org/archives/list/cockpit-devel@lists.fedorahosted.org/thread/EC7YXZPIWCBDUNOK6336YRRTJE62RHU2/), this error message is terrible. Let's fix that.

**Comment by martinpitt** [comment]
After `dnf remove sassc`, I get an error which is not quite as bad:
```
    Error: spawnSync sassc ENOENT
        at Object.spawnSync (internal/child_process.js:1067:20)
        at spawnSync (child_process.js:602:24)
        at Object.execFileSync (child_process.js:629:15)
        at Object.module.exports (/var/home/martin/upstream/starter-kit/src/lib/sassc-loader.js:12:18)
```
It's still a bit hard to read, though.

**Comment by martinpitt** [comment]
Sent fix in PR #404
